SYCL/DeviceArchitecture/device_architecture.cpp

```diff
@@ -0,0 +1,70 @@
+// REQUIRES: opencl-aot, ocloc, cpu, gpu
+// UNSUPPORTED: cuda
+// CUDA is not supported by sycl_ext_intel_device_architecture extension.
+
+// TODO: remove w/a with "-D__SYCL_TARGET_INTEL_X86_64__
+// -D__SYCL_TARGET_INTEL_GPU_SKL__" after bug in compiler driver related to
+// if_architecture_is feature is resolved
+// RUN: %clangxx -fsycl -fsycl-targets=intel_gpu_skl,spir64_x86_64 -D__SYCL_TARGET_INTEL_X86_64__ -D__SYCL_TARGET_INTEL_GPU_SKL__ %s -o %t.out
+// RUN: %t.out
+
+#include <sycl/ext/intel/experimental/device_architecture.hpp>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::intel::experimental;
+
+int main() {
+  std::vector<int> vec(4);
+  {
+    buffer<int> buf(vec.data(), vec.size());
+
+    queue gpu_queue(gpu_selector_v);
+
+    gpu_queue.submit([&](handler &cgh) {
+      auto acc = buf.get_access<access::mode::read_write>(cgh);
+      cgh.single_task([=]() {
+        if_architecture_is<architecture::intel_gpu_skl>([&]() {
+          acc[0] = 1;
+        }).otherwise([&]() { acc[0] = 0; });
+      });
+    });
+
+    gpu_queue.submit([&](handler &cgh) {
+      auto acc = buf.get_access<access::mode::read_write>(cgh);
+      cgh.single_task([=]() {
+        if_architecture_is<architecture::intel_gpu_dg1>([&]() {
+          acc[1] = 0;
+        }).otherwise([&]() { acc[1] = 2; });
+      });
+    });
+
+    gpu_queue.submit([&](handler &cgh) {
+      auto acc = buf.get_access<access::mode::read_write>(cgh);
+      cgh.single_task([=]() {
+        if_architecture_is<architecture::intel_gpu_dg1>([&]() { acc[2] = 0; })
+            .else_if_architecture_is<architecture::intel_gpu_9_0_9>(
+                [&]() { acc[2] = 3; })
+            .otherwise([&]() { acc[2] = 0; });
+      });
+    });
+
+    queue cpu_queue(cpu_selector_v);
+
+    cpu_queue.submit([&](handler &cgh) {
+      auto acc = buf.get_access<access::mode::read_write>(cgh);
+      cgh.single_task([=]() {
+        if_architecture_is<architecture::x86_64>([&]() {
+          acc[3] = 4;
+        }).otherwise([&]() { acc[3] = 0; });
+      });
+    });
+  }
+
```

Comment:
I thought between the GitHub Actions and the Jenkins tests, that llvm-test-suite ends up running on a couple of different machines (CoffeeLake, SkyLake). Also, while not part of our regular CI tests, QA runs them against SandyBridge and others from time to time, if I'm not mistaken.  Won't that change the values and cause these asserts to fail?

Reply:
For this test to be reliable, it should require certain hardware. It looks like such information should be available through environment for us:

https://github.com/intel/llvm-test-suite/blob/d249d72361da11aee3adc992bbed581bf2d6ecaf/SYCL/lit.cfg.py#L103-L109

Therefore, I suggest that we limit this test to a particular HW. For more extensive testing, we could go with unit-tests: we don't really need a device to check if we took one code path or another based on presence of a macro

Reply:
Made the test dependent on CPU only. Logic for GPU platforms is testing in unit tests